apps/aeoracle/src/aeo_state_tree.erl

```diff
@@ -30,32 +30,45 @@
 %%% Types
 %%%===================================================================
 
--type mkey() :: aeo_oracles:id() | aeo_interaction:id().
--type mvalue() :: aeo_oracles:serialized() | aeo_interaction:serialized().
--type mtree() :: aeu_mtrees:tree(mkey(), mvalue()).
+%% The main tree contains oracles - then there is a set of trees (one per
+%% oracle) that contains the interactions... The 'oracle' field only contain
+%% the root hash of the interaction tree.
+%%
+%% We also keep a cache of objects to expire.
+
+-type otree() :: aeu_mtrees:tree(aeo_oracles:id(), aeo_oracles:serialized()).
+-type itree() :: aeu_mtrees:tree(aeo_interaction:id(), aeo_interaction:serialized()).
 -type interaction() :: aeo_interaction:interaction().
 -type oracle() :: aeo_oracles:oracle().
--type cache() :: gb_sets:set({integer(), binary()}).
+-type itrees() :: gb_tree:tree(aeo_oracles:id(), itree()).
+-type cache_item() :: {oracle, aeo_oracles:id()}
+                    | {interaction, aeo_oracles:id(), aeo_interaction:id()}.
+-type cache() :: gb_sets:set({integer(), cache_item()}).
 -type block_height() :: non_neg_integer().
 
--record(oracle_tree, { mtree = gb_merkle_trees:empty() :: mtree()
-                     , cache = cache_new() :: cache()
+-record(oracle_tree, { otree  = aeu_mtrees:empty() :: otree()
+                     , itrees = gb_trees:empty()   :: itrees()
```

Comment:
Not a merkle tree? I am trying to understand how it works.

Reply:
I see - so each oracle has an oracle interaction tree, though the oracle internal representation does not contain the oracle interaction tree (as I was expecting) instead it contains its root hash. Then the oracle interactions trees - one per oracle - in the internal representation are stored completely separately, and the consistency between the two is kept with additional code.

I admit this is not very clear from the comment https://github.com/aeternity/epoch/pull/572/files/7d61e16e9d72e5dd8e6a9ddeff10b940c748c1c8#diff-e72d14988d10e82668a1a08726718bb8R33

Reply:
I've rephrased the comment - and clarified that the reason for having the interactions in a separate tree is to avoid lots of serialize/deserialize on insert/update.